rosbag2/test/rosbag2/rosbag2_write_integration_fixture.hpp

```diff
@@ -91,20 +87,50 @@ class RosBag2IntegrationTestFixture : public Rosbag2TestFixture
     std::shared_ptr<rosbag2_storage::SerializedBagMessage> message,
     std::string topic_name, size_t number_expected_messages)
   {
-    messages_stored_counter_.insert({topic_name, 0});
     publisher_futures_.push_back(std::async(
         std::launch::async, [this, message, topic_name, number_expected_messages]() {
           auto publisher = publisher_node_->create_publisher<std_msgs::msg::String>(topic_name);
 
-          while (rclcpp::ok() && messages_stored_counter_[topic_name] < number_expected_messages) {
-            publisher->publish(message->serialized_data);
-            // We need to wait a bit, in order for record to process the messages
+          rosbag2_storage_plugins::SqliteWrapper
+          db(database_name_, rosbag2_storage::storage_interfaces::IOFlag::READ_ONLY);
+          while (rclcpp::ok() && count_stored_messages(db, topic_name) < number_expected_messages) {
+            publisher->publish(message->serialized_data.get());
+            // rate limiting
             std::this_thread::sleep_for(50ms);
           }
         }
     ));
   }
 
+  size_t count_stored_messages(rosbag2_storage_plugins::SqliteWrapper & db, std::string topic_name)
+  {
+    // protect against concurrent writes (from recording) that may make the count query throw.
+    while (true) {
+      try {
+        return count_stored_messages_in_db(db, topic_name);
+      } catch (const rosbag2_storage_plugins::SqliteException & e) {
+        (void) e;
```

Comment:
shouldn't this break the while loop? Is there a way to have the loop conditioned otherwise than `while (true)`? This seems just dangerous.

Reply:
The test runner already provides a timeout so the `while (true)` while never run forever.
If a query fails to lock the sqlite db the wrapper throws (see also comment below). This loop simply retries the query until it succeeds (or the test wrapper kills and fails the test). It could be rewritten but this would only hide the intention of looping until the query succeeds.